integration-tests/src/tests/client/flat_storage.rs

```diff
@@ -123,46 +163,191 @@ fn test_flat_storage_creation() {
         })
     );
 
-    // Run chain for a couple of blocks and check that statuses switch to `CatchingUp` and then to `Ready`.
-    // State is being fetched in rayon threads, but we expect it to finish in <30s because state is small and there is
-    // only one state part.
-    const BLOCKS_TIMEOUT: BlockHeight = 30;
-    let start_height = 8;
-    let mut next_height = start_height;
-    let mut was_catching_up = false;
-    while next_height < start_height + BLOCKS_TIMEOUT {
-        env.produce_block(0, next_height);
-        env.clients[0].run_flat_storage_creation_step().unwrap();
-        next_height += 1;
-        match store_helper::get_flat_storage_state_status(&store, 0) {
-            FlatStorageStateStatus::FetchingState(..) => {
-                assert!(!was_catching_up, "Flat storage state status inconsistency: it was catching up before fetching state");
-            }
-            FlatStorageStateStatus::CatchingUp => {
-                was_catching_up = true;
-            }
-            FlatStorageStateStatus::Ready => {
-                assert!(
-                    was_catching_up,
-                    "Flat storage state is ready but there was no flat storage catchup observed"
+    wait_for_flat_storage_creation(&mut env, 8);
+}
+
+/// Check that client can create flat storage on some shard while it already exists on another shard.
+#[test]
+fn test_flat_storage_creation_two_shards() {
+    init_test_logger();
+    let num_shards: NumShards = 2;
+    let genesis = Genesis::test_sharded_new_version(
+        vec!["test0".parse().unwrap()],
+        1,
+        vec![1; num_shards as usize],
+    );
+    let chain_genesis = ChainGenesis::new(&genesis);
+    let store = create_test_store();
+
+    // Process some blocks with flat storages for two shards. Then remove flat storage data from disk for shard 0.
+    {
+        let runtimes: Vec<Arc<dyn RuntimeAdapter>> = vec![Arc::new(
+            nearcore::NightshadeRuntime::test(Path::new("../../../.."), store.clone(), &genesis),
+        )];
+        let mut env =
+            TestEnv::builder(chain_genesis.clone()).runtime_adapters(runtimes.clone()).build();
```

Comment:
This seems to be a common pattern to create `env` copy&pasted thought this file. Maybe let's extract it to a separate method, especially considering that `runtimes` only used to create `env` (which also makes `.clone()` redundant).